### PR TITLE
Async Listener Stop

### PR DESCRIPTION
**Pull request by nibanks**
Another suggestion from #1709. cc @geoffkizer

### CODE REVIEW

File: src/platform/platform_posix.c [hunk 1]
```diff
@@ -270,22 +270,23 @@ CxPlatRefIncrement(
 
 BOOLEAN
 CxPlatRefIncrementNonZero(
-    _Inout_ volatile CXPLAT_REF_COUNT* RefCount
+    _Inout_ volatile CXPLAT_REF_COUNT* RefCount,
+    _In_ uint32_t Bias
```
Comment (anrossi) [inline src/platform/platform_posix.c:274]: For adding more than 1 ref at a time?

**Review by anrossi — CHANGES_REQUESTED** [review]
Update docs with the new event, and add tests for calling `ListenerStop` on the worker thread, and the app thread, some API tests to validate calling Stop before Start fails, etc.

File: src/platform/platform_posix.c (hunk 1 above)
Comment (nibanks) [reply src/platform/platform_posix.c:274]: This particular change was just to make all platforms have the same signature.

File: src/core/listener.c [hunk 2]
```diff
@@ -122,27 +155,28 @@ MsQuicListenerClose(
...
 
-    //
-    // Make sure the listener has unregistered from the binding.
-    //
-    MsQuicListenerStop(Handle);
+    QUIC_LIB_VERIFY(!Listener->AppClosed);
+    Listener->AppClosed = TRUE;
 
-    CxPlatRundownUninitialize(&Listener->Rundown);
+    if (Listener->StopCompleteThreadID == CxPlatCurThreadID()) {
```
Comment (thhous-msft) [inline src/core/listener.c:162]: I'm not quite sure on this. What happens if the app calls ListenerClose() from a different callback?

Comment (nibanks) [reply src/core/listener.c:162]: There's nothing we can do unless really.

File: docs/api/ListenerClose.md [hunk 3]
```diff
@@ -23,7 +23,7 @@ A valid handle to an open listener object.
 
 # Remarks
 
-`ListenerClose` frees all allocated resources associated with the listener handle. If a listener has not had [ListenerStop](ListenerStop.md) called on it at the time `ListenerClose` is called, [ListenerStop](ListenerStop.md) is invoked internally. Due to this, a server application **MUST NOT** call `ListenerClose` within any callback unless it has already called [ListenerStop](ListenerStop.md) outside of any callbacks.
+`ListenerClose` frees all allocated resources associated with the listener handle. If a listener has not had [ListenerStop](ListenerStop.md) called on it at the time `ListenerClose` is called, [ListenerStop](ListenerStop.md) is invoked internally. `ListenerClose` may be called on any Listener callback.
```
Comment (thhous-msft) [inline docs/api/ListenerClose.md:26]: This needs to be changed that ListenerClose can only be called from another thread or the StopComplete event.

**Review by thhous-msft — DISMISSED** [review]

File: docs/api/ListenerClose.md [hunk 4]
```diff
@@ -23,7 +23,7 @@ A valid handle to an open listener object.
 
 # Remarks
 
-`ListenerClose` frees all allocated resources associated with the listener handle. If a listener has not had [ListenerStop](ListenerStop.md) called on it at the time `ListenerClose` is called, [ListenerStop](ListenerStop.md) is invoked internally. Due to this, a server application **MUST NOT** call `ListenerClose` within any callback unless it has already called [ListenerStop](ListenerStop.md) outside of any callbacks.
+`ListenerClose` frees all allocated resources associated with the listener handle. If a listener has not had [ListenerStop](ListenerStop.md) called on it at the time `ListenerClose` is called, [ListenerStop](ListenerStop.md) is invoked internally. `ListenerClose` may be called on any Listener callback. A server application **MUST NOT** call `ListenerClose` within any callback except for `QUIC_LISTENER_EVENT_STOP_COMPLETE`, or unless it has already received the `QUIC_LISTENER_EVENT_STOP_COMPLETE` event.
```
Comment (anrossi) [inline docs/api/ListenerClose.md:26]: The second part here seems to contradict the first. It says you can call `ListenerClose` without calling `ListenerStop` first. But then it says you cannot call `ListenerClose` unless you have received the `QUIC_LISTENER_EVENT_STOP_COMPLETE` event.

**Review by anrossi — APPROVED** [review]

File: docs/api/ListenerClose.md (hunk 4 above)
Comment (nibanks) [reply docs/api/ListenerClose.md]: Fixed

**Review by thhous-msft — APPROVED** [review]

**Pull request merged**